### PR TITLE
Bug in mediabox expansion when applying non-right angle rotation

### LINKED ISSUE #2281: Bug in mediabox expansion when applying non-right angle rotation
- I have a pdf that I want to rotate by an angle that is not a multiple of 90. Let's say 175 degrees.
- In addition, I want to update the `mediabox` coordinates as needed after the rotation.

**Currently, the `mediabox` expansion is broken.**

The bug is caused by these lines of code: https://github.com/py-pdf/pypdf/blob/9047079045742f09df1a2f5ff40d4b316ec35f7a/pypdf/_page.py#L1658-L1662

## Environment

Which environment were you using when you encountered the problem?

```bash
$ python -m platform
macOS-14.0-x86_64-i386-64bit

$ python -c "import pypdf;print(pypdf._debug_versions)"
pypdf==3.15.5, crypt_provider=('cryptography', '39.0.1'), PIL=10.0.1
```

## Code + PDF

This is a minimal, complete example that shows the issue:

```python
from pypdf import PdfWriter, PdfReader, Transformation

reader = PdfReader("input.pdf")
writer = PdfWriter()

transformation = Transformation().rotate(175)

for page_box in reader.pages:
    page_box.add_transformation(transformation, expand=True)
    writer.add_page(page_box)

with open("output.pdf", "wb") as f:
    writer.write(f)
```

- Input pdf: [input.pdf](https://github.com/py-pdf/pypdf/files/13219150/input.pdf)
- Current output of the code above: [current_output.pdf](https://github.com/py-pdf/pypdf/files/13219155/current_output.pdf)
  - Notice the white space covering the top-half of each page. This represents an incorrect expansion of the media box.
- Expected output of the code above: [expected_output.pdf](https://github.com/py-pdf/pypdf/files/13219165/expected_output.pdf)

### PR DESCRIPTION
**Pull request by MrinalJain17**
Fixes #2281

### CODE REVIEW

**Comment by pubpub-zz** [comment]
@MrinalJain17, can you confirm that when applying angle from -370 to +360, all results are as expected ?

**Comment by MrinalJain17** [comment]
Yes @pubpub-zz 

In fact, given below is a comparison of the dimensions of a page after it has been rotated with `pypdf` vs `pillow`. With expansion enabled in both scenarios.

I've also manually checked the results of course.

Current (before fix):

- Note the height discrepancy when rotated by an angle of 175 degrees aligns with the sample I attached here #2281 

image

New (after fix):

278074508-ea25073a-d536-489e-9e39-bfa29578adc6-2

**Comment by pubpub-zz** [comment]
your results are very good! can you add a test in order to make sure that the issue will not come back ?

File: tests/test_page.py [hunk 1]
```diff
@@ -122,6 +122,28 @@ def test_page_operations(pdf_path, password):
...
+        writer.add_page(page_box)
+
+    rotated_pdf_bytes = BytesIO()
+    writer.write_stream(rotated_pdf_bytes)
+
+    expected_pdf_path = RESOURCE_ROOT / "crazyones_rotated_175.pdf"
+    with open(expected_pdf_path, "rb") as f:
+        expected_pdf_bytes = BytesIO(f.read())
+
+    assert rotated_pdf_bytes.getvalue() == expected_pdf_bytes.getvalue()
```
Comment (pubpub-zz) [inline tests/test_page.py:144]: I would have just look at the box size as you did in your graphs with a criteria <1E-4

Comment (MrinalJain17) [reply tests/test_page.py]: That makes sense @pubpub-zz . I've updated the tests.

Due to rounding/interpolation, there's deviation of about 1-ish pixels compared to pillow. I've set an absolute tolerance of 2 pixels instead.

Comment (MrinalJain17) [reply tests/test_page.py]: Lmk if this looks good. Or if you want me to set the expected dimensions to be what `pypdf` actually ends up with -- and compare with a small tolerance instead.

**Comment by MrinalJain17** [comment]
Hi @pubpub-zz. Checking in on this -- lmk if there's anything else needed.

**Comment by pubpub-zz** [comment]
This looks good to me👏
@MartinThoma all yours😉

File: tests/test_page.py [hunk 2]
```diff
@@ -122,6 +123,33 @@ def test_page_operations(pdf_path, password):
...
+    [
+        (175, 680, 844),
+        (45, 994, 994),
+        (-80, 888, 742),
+    ]
+)
+def test_mediabox_expansion_after_rotation(angle: float, expected_width: int, expected_height: int):
+    """
+    This test validates the dimensions of a pdf's mediabox after applying rotation with
+    expansion (at a non-right angle).
```
Comment (MartinThoma) [inline tests/test_page.py:137]:
```suggestion
    """
    Mediabox dimensions after rotation at a non-right angle with expension are correct.

    The test was validated against pillow (see PR #2282)
    """
```

File: tests/test_page.py [hunk 3]
```diff
@@ -122,6 +123,35 @@ def test_page_operations(pdf_path, password):
...
+        (-80, 888, 742),
+    ]
+)
+def test_mediabox_expansion_after_rotation(angle: float, expected_width: int, expected_height: int):
+    """
+    Mediabox dimensions after rotation at a non-right angle with expension are correct.
+
+    The test was validated against pillow (see PR #2282)
+    """
+    """
```
Comment (MartinThoma) [inline tests/test_page.py:140]:
```suggestion
```

**Comment by MartinThoma** [comment]
@MrinalJain17 Thank you for this fix and the explanation around it in the PR :pray: 

If you want, I can add you to https://pypdf.readthedocs.io/en/latest/meta/CONTRIBUTORS.html :-)

**Pull request merged**